gdk-pixbuf: multiple module files

cwyc opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
The gdk-pixbuf library can search for "modules" or "loaders" to extend its ability to load image formats. The modules available on a system are gathered and listed in a `loaders.cache` file, which gdk-pixbuf looks for in `$GDK_PIXBUF_MODULE_FILE`. 

Gdk-pixbuf recognizes only one loaders.cache file. This isn't a problem on other distributions, which call `gdk-pixbuf-query-loaders` after installation to update the systemwide loaders.cache that contains every installed loader.

But nix can't do that, so 1) a program can't be packaged with loaders from multiple packages, and 2) if a program is packaged with a loader, the user can't extend it with other loaders in their environment.

We see the first issue with pretty much all GTK apps, which need loaders from librsvg as well as gdk-pixbuf. Our [solution](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/librsvg/default.nix#L115) is to manually merge all of gdk-pixbuf's loaders.cache entries in to librsvg's, and use only the latter. But that doesn't support any other package providing loaders.

###### Things done
A better way could be for gdk-pixbuf to accept multiple loaders.cache. (see https://github.com/NixOS/nixpkgs/pull/42562#pullrequestreview-131597125 )

First, module loading code in gdk-pixbuf is patched to read multiple paths from `$GDK_PIXBUF_MODULE_FILE`, separated by a colon.

Second, `wrapGappsHook`, which sets `$GDK_PIXBUF_MODULE_FILE` for most executables using gdk-pixbuf, is patched so that it appends to it rather than replaces it. This will let user-set loaders in `services.xserver.gdk-pixbuf.modulePackages` be recognized.

Third, the setup hook by which `gdk-pixbuf` includes its own loaders is updated to take advantage of this feature. (see https://github.com/NixOS/nixpkgs/pull/102189#issuecomment-719919574 )

Currently, packages include the hybrid-librsvg-gdk-pixbuf-loaders.cache either by setting `$GDK_PIXBUF_MODULE_FILE` for `wrapGappsHook` to pick up, or by manually supplying it to `makeWrapper`.
Even though it's possible to supply librsvg and gdk-pixbuf separately, I've left librsvg as is for compatibility's sake.
And while this patches the `wrapGappsHook` packages to append rather than replace the environment variable, the manually-wrapped packages are unchanged.

Warning: this triggers a lot of compilation.

Documentation hasn't been updated.

Tagging @jtojnar

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] TODO Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] TODO (Package updates) Added a release notes entry if the change is major or breaking 
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cwyc committed:
gdk-pixbuf: multiple module files
patches gdk-pixbuf to accept multiple module files in $GDK_PIXBUF_MODULE_FILE, separated by ":"

jtojnar commented:
Sorry for the late reply, this got lost in my queue. I think this is reasonable compromise between performance and convenience for packaging.

Primary concerns would be that gdk-pixbuf does not normally allow this so I am not sure how well does [`gdk_pixbuf_io_init_modules` function](https://gitlab.gnome.org/GNOME/gdk-pixbuf/-/blob/ab562505fb12887dd7a1a4e6a467087e76e036ca/gdk-pixbuf/gdk-pixbuf-io.c#L378) handle being called multiple times, potentially with overlapping cache files.

Some basic tests along the line of https://github.com/NixOS/nixpkgs/blob/b77494badb8995a024b72afccc597fffac6bc041/pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix#L40 would go a long way in making sure this does not bring any unfortunate breakage, without having to rebuild the world.

jtojnar commented on pkgs/development/libraries/gdk-pixbuf/setup-hook.sh:
```diff
@@ -1,12 +1,9 @@
         if [[ -f "${GDK_PIXBUF_MODULE_FILE-}" ]]; then
-            if (( "$(cat "$loadersCache" | wc -l)" > "$(cat "$GDK_PIXBUF_MODULE_FILE" | wc -l)" )); then
-                export GDK_PIXBUF_MODULE_FILE="$loadersCache"
-            fi
+            export GDK_PIXBUF_MODULE_FILE="$loadersCache":$GDK_PIXBUF_MODULE_FILE
         else
             export GDK_PIXBUF_MODULE_FILE="$loadersCache"
         fi
```
This will clear the variable e.g. when `GDK_PIXBUF_MODULE_FILE` already contains multiple values. Just switch it to

```
addToSearchPath GDK_PIXBUF_MODULE_FILE "$loadersCache"
```

jtojnar commented on pkgs/build-support/setup-hooks/wrap-gapps-hook/wrap-gapps-hook.sh:
```diff
@@ -11,7 +11,7 @@ addEnvHooks "${targetOffset:?}" find_gio_modules
 gappsWrapperArgsHook() {
     if [ -n "$GDK_PIXBUF_MODULE_FILE" ]; then
-        gappsWrapperArgs+=(--set GDK_PIXBUF_MODULE_FILE "$GDK_PIXBUF_MODULE_FILE")
+        gappsWrapperArgs+=(--suffix GDK_PIXBUF_MODULE_FILE : "$GDK_PIXBUF_MODULE_FILE")
```
I wonder if `--prefix` might be safer here. E.g. in case user has `GDK_PIXBUF_MODULE_FILE` in their environment containing ABI incompatible loaders (assuming the cache format is stable, first registered module for given file takes precedence, and modules are loaded lazily). But it is not clear to me what handles matching the image to a module, so :woman_shrugging:

Artturin commented on pkgs/development/libraries/gdk-pixbuf/default.nix:
```diff
@@ -42,6 +42,8 @@ stdenv.mkDerivation rec {
     # Move installed tests to a separate output
     ./installed-tests-path.patch
+    # Allow for multiple loader.cache files specified in $GDK_PIXBUF_MODULE_FILE, delimited by ":"
+    ./multiple-module-files.patch
```
Could this be submitted upstream? That way we could get some upstream reviews for it.

jtojnar replied:
We can try, but I doubt upstream will be very happy about this. The environment variable is mostly meant for development and what we are doing is an extremely dirty hack.

cwyc commented:
I share your concern about the robustness of the patch. With some tests, hopefully we'll have a better idea of its vulnerabilities.
Though it seems to me like the code is somewhat resilient to loading multiple files. The module loading procedure loads a number of [builtin modules](https://gitlab.gnome.org/GNOME/gdk-pixbuf/-/blob/master/gdk-pixbuf/gdk-pixbuf-io.c#L673) before reading `loaders.cache`. And in [`gdk_pixbuf_io_init_modules`](https://gitlab.gnome.org/GNOME/gdk-pixbuf/-/blob/master/gdk-pixbuf/gdk-pixbuf-io.c#L412), it expects modules to possibly already exist in `file_formats` prior to invocation. 
Of course, upstream still assumes a single loaders.cache, so the internal interfaces can change at any time.

---

About --suffix, I think (?) later entries in $GDK_PIXBUF_MODULE_FILE have precedence.
The loop calls [gdk_pixbuf_io_init_modules](https://gitlab.gnome.org/GNOME/gdk-pixbuf/-/blob/master/gdk-pixbuf/gdk-pixbuf-io.c#L378) for each file in order, which then [prepends](https://gitlab.gnome.org/GNOME/gdk-pixbuf/-/blob/master/gdk-pixbuf/gdk-pixbuf-io.c#L424) loaders from the file to [file_formats](https://gitlab.gnome.org/GNOME/gdk-pixbuf/-/blob/master/gdk-pixbuf/gdk-pixbuf-io.c#L169). So the order in file_formats is reversed from that of $GDK_PIXBUF_MODULE_FILE. 
[_gdk_pixbuf_get_module](https://gitlab.gnome.org/GNOME/gdk-pixbuf/-/blob/master/gdk-pixbuf/gdk-pixbuf-io.c#L885) then selects which module to load by iterating through file_formats in order. 
If it's built with GDK_PIXBUF_USE_GIO_MIME, it [selects](https://gitlab.gnome.org/GNOME/gdk-pixbuf/-/blob/master/gdk-pixbuf/gdk-pixbuf-io.c#L918) the first module in file_formats with a matching MIME type.
If it's built without MIME support, then it searches the entire list for the module with the best ["score"](https://gitlab.gnome.org/GNOME/gdk-pixbuf/-/blob/master/gdk-pixbuf/gdk-pixbuf-io.c#L944). [Comparison](https://gitlab.gnome.org/GNOME/gdk-pixbuf/-/blob/master/gdk-pixbuf/gdk-pixbuf-io.c#L945) is done with a > rather than a >=, so if multiple modules provide the highest score, the earliest encountered will remain the selected maximum.

In any case, the opposite precedence behavior seems to be the standard, and I agree modules included with the package should have precedence. So I'll change it to --prefix, and edit the patch to process the path in reverse order.

---

I wonder if packaging procedures will have to change.
I can think of these behaviours that packages rely on:
- when gdk-pixbuf or librsvg is added to buildInputs, its setupHook loads its loaders.cache to the build environment $GDK_PIXBUF_MODULE_FILE
- when both are added to buildInputs, their setupHooks work out the conflict, and the functionality of both appears
- wrapGAppsHook takes $GDK_PIXBUF_MODULE_FILE in the build environment as what it should include with the executable.
	- There may be an expectation that including the package in buildInputs is all that's needed to gain loader support, disregarding the role of wrapGAppsHook
- the librsvg loaders.cache contains the standard gdk-pixbuf loaders as well.
	- Though apps typically include both as dependencies, when manually constructing a wrapper, only librsvg's loaders.cache is set.
- obviously, the library should work with only one loaders.cache in $GDK_PIXBUF_MODULE_FILE

Seeing how many gtk programs there are, I imagine they behaviours have to stay. But it doesn't really indicate how we should design the interface for packaging more-than-one loaders.cache.

Should we make all loader packages have setupHooks that append to $GDK_PIXBUF_MODULE_FILE, just like shared libraries do to $LD_LOADERS_PATH? But unlike shared libraries, modules often overlap: do we then need a way to specify their order?

Though the vast majority of programs just need the standard librsvg/gdk-pixbuf to render GTK icons. Perhaps wrapGAppsHook can include them automatically without the setupHooks, packages not directly calling gdk-pixbuf no longer need to explicitly include them in buildInputs (but it doesn't hurt), and packages needing other loaders can manually wrap them (maintaining the librsvg-containing-gdk-pixbuf behaviour).

cwyc committed:
reverse module loading order
(will squash before committing)

cwyc committed:
modified build support code
(will squash before committing)

jtojnar commented on doc/languages-frameworks/gnome.section.md:
```diff
@@ -92,7 +92,7 @@ For convenience, it also adds `dconf.lib` for a GIO module implementing a GSetti
 - []{#ssec-gnome-hooks-glib} `glib` setup hook will populate `GSETTINGS_SCHEMAS_PATH` and then `wrapGAppsHook` will prepend it to `XDG_DATA_DIRS`.
 
-- []{#ssec-gnome-hooks-gdk-pixbuf} `gdk-pixbuf` setup hook will populate `GDK_PIXBUF_MODULE_FILE` with the path to biggest `loaders.cache` file from the dependencies containing [GdkPixbuf loaders](ssec-gnome-gdk-pixbuf-loaders). This works fine when there are only two packages containing loaders (`gdk-pixbuf` and e.g. `librsvg`) – it will choose the second one, reasonably expecting that it will b ...
+- []{#ssec-gnome-hooks-gdk-pixbuf} `wrapGAppsHook` will automatically prepend `GDK_PIXBUF_MODULE_FILE` with modules from `gdk-pixbuf` and `librsvg`. To include additional gdk-pixbuf modules, set `extraGdkPixbufModules` with an array of packages.
```
This section lists related setup hooks, this should probably go elsewhere.

cwyc replied:
I don't quite understand, where is is better?

jtojnar commented on pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix:
```diff
@@ -36,7 +37,12 @@ makeSetupHook {
   ];
-  substitutions = {
+  substitutions = let
+    loadersPath = (lib.strings.removeSuffix "/" gdk-pixbuf.moduleDir) + ".cache";
```
We might want to expose `cacheFile` in the `gdk-pixbuf.passthru`.

jtojnar commented on pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix:
```diff
@@ -36,7 +37,12 @@ makeSetupHook {
+  substitutions = let
+    loadersPath = (lib.strings.removeSuffix "/" gdk-pixbuf.moduleDir) + ".cache";
+  in {
+    STANDARD_GDK_PIXBUF_MODULES = "${gdk-pixbuf}/${loadersPath}:${librsvg}/${loadersPath}";
```
Then we will able to use `lib.makeSearchPathOutput "lib" gdk-pixbuf.cacheFile [ gdk-pixbuf librsvg ]`.

jtojnar commented on pkgs/development/libraries/gdk-pixbuf/default.nix:
```diff
@@ -42,6 +42,8 @@ stdenv.mkDerivation rec {
   patches = [
     # Move installed tests to a separate output
     ./installed-tests-path.patch
+    # Allow for multiple loader.cache files specified in $GDK_PIXBUF_MODULE_FILE, delimited by ":"
```
loader → loaders

cwyc committed:
more changes
will be squashed before merge

cwyc committed:
tests for wrapper code
to do: tests for patched code
will be squashed before merge

jtojnar commented on doc/languages-frameworks/gnome.section.md:
```diff
@@ -34,11 +34,11 @@ To allow software to use various virtual file systems, `gvfs` package can be als
+Each loader package will contain a `lib/gdk-pixbuf-2.0/2.10.0/loaders.cache` file containing information about the available loaders. GdkPixbuf looks for these files in the `GDK_PIXBUF_MODULE_FILE` environment variable. Although upstream GdkPixbuf only supports a single file, in nixpkgs it is patched to accept multiple files, separated by `:`.
 
-`gdk-pixbuf` contains [a setup hook](#ssec-gnome-hooks-gdk-pixbuf) that sets `GDK_PIXBUF_MODULE_FILE` from dependencies but as mentioned in further section, it is pretty limited. Loaders should propagate this setup hook.
+[`wrapGAppsHook`]{#ssec-gnome-hooks-wrapgappshook} handles setting `GDK_PIXBUF_MODULE_FILE` for GTK apps. If you're manually constructing a wrapper, consider `--prefix`ing the variable rather than overwriting it, so that users can extend your program's capabilities through `services.xserver.gdk-pixbuf.modulePackages`.
```
`{#ssec-gnome-hooks-wrapgappshook}` (with curly braces) adds an anchor, so this crates an anchor that already exists, making the manual build to fail.

jtojnar commented on pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix:
```diff
@@ -64,6 +68,29 @@ makeSetupHook {
+        '';
+      };
+
+      pixbuf-modules = basic.override {extraGdkPixbufModules = [mock-pixbuf-module];};
```
This should be `overrideAttrs`?

cwyc committed:
corrected mistakes
will be squashed before merging

cwyc committed:
added package test for gdk-pixbuf multiple file functionality
will be squashed before merging

cwyc commented:
Sorry for the delay. I've made some corrections, and added a simple test for gdk-pixbuf.

cwyc commented:
Still need to modify [`services.xserver.gdk-pixbuf.modulePackages`](https://github.com/NixOS/nixpkgs/blob/nixos-21.11/nixos/modules/services/x11/gdk-pixbuf.nix) module, as it no longer needs to generate an amalgamated loaders.cache, and add tests for it.

jtojnar commented on pkgs/build-support/setup-hooks/wrap-gapps-hook/wrap-gapps-hook.sh:
```diff
@@ -10,8 +10,13 @@ find_gio_modules() {
+
+    if [ -n "$extraGdkPixbufModules" ]; then
+        for pkg in $extraGdkPixbufModules; do
+            gappsWrapperArgs+=(--prefix GDK_PIXBUF_MODULE_FILE : "$pkg/@gdkPixbufCacheFile@")
```
I imagine we would create two dummy loaders producing images of different colours like the following to test that the priorities are correct (`extraGdkPixbufModules` takes priority over both `standardGdkPixbufModules` and `GDK_PIXBUF_MODULE_FILE` from environment).

```c
#include <gdk-pixbuf/gdk-pixbuf-io.h>


#ifndef
#define PIXBUF_COLOR "#ff0000"
#endif


G_MODULE_EXPORT void fill_vtable(GdkPixbufModule* module);

G_MODULE_EXPORT void fill_info(GdkPixbufFormat* info);


typedef struct {
    GdkPixbufModuleUpdatedFunc update_func;
    GdkPixbufModulePreparedFunc prepare_func;
    GdkPixbufModuleSizeFunc size_func;
    gpointer user_data;
} TestPixbufCtx;


static gpointer begin_load(
    GdkPixbufModuleSizeFunc size_func,
    GdkPixbufModulePreparedFunc prepare_func,
    GdkPixbufModuleUpdatedFunc update_func,
    gpointer user_data,
    GError** error
) {
    TestPixbufCtx* tpc;

    tpc = g_new0(TestPixbufCtx, 1);
    tpc->size_func = size_func;
    tpc->prepare_func = prepare_func;
    tpc->update_func = update_func;
    tpc->user_data = user_data;

    return tpc;
}


static char * xpm[] = {
    // width, height, #colors, #chars_per_pixel
    "1 1 1 1",
    // palette
    "a c " PIXBUF_COLOR,
    // image
    "a"
};

static gboolean stop_load(gpointer context, GError** error) {
    TestPixbufCtx* tpc = (TestPixbufCtx*) context;
    int width, height;
    int requested_width, requested_height;
    g_autoptr(GdkPixbuf) pixbuf;


    width = 1;
    height = 1;
    requested_width = width;
    requested_height = height;

    if (tpc->size_func) {
        (*tpc->size_func)(&requested_width, &requested_height, tpc->user_data);
    }

    pixbuf = gdk_pixbuf_new_from_xpm(xpm);

    if (tpc->prepare_func) {
        (*tpc->prepare_func)(pixbuf, NULL, tpc->user_data);
    }

    if (tpc->update_func != NULL) {
        (*tpc->update_func)(pixbuf, 0, 0, gdk_pixbuf_get_width(pixbuf), gdk_pixbuf_get_height(pixbuf), tpc->user_data);
    }

    g_clear_object(&pixbuf);
    g_free(tpc);

    return TRUE;
}


static gboolean load_increment(gpointer context, const guchar* buf, guint size, GError** error) {
    return TRUE;
}


void fill_vtable(GdkPixbufModule* module) {
    module->begin_load = begin_load;
    module->stop_load = stop_load;
    module->load_increment = load_increment;
}


void fill_info(GdkPixbufFormat* info) {
    static GdkPixbufModulePattern signature[] = {
        {"GIF8", "", 100},
        {NULL, NULL, 0}
    };

    static gchar* mime_types[] = {
        "image/gif",
        NULL
    };

    static gchar* extensions[] = {
        "gif",
        NULL
    };

    info->name = "test";
    info->signature = signature;
    info->domain = NULL;
    info->description = "Loader for testing priority of loaders";
    info->mime_types = mime_types;
    info->extensions = extensions;
    info->flags = GDK_PIXBUF_FORMAT_THREADSAFE;
    info->disabled = FALSE;
    info->license = "MIT";
}
```

jtojnar commented on nixos/doc/manual/release-notes/rl-2205.section.md:
```diff
@@ -621,6 +621,17 @@ In addition to numerous new and upgraded packages, this release has the followin
+  - Manually wrapping `$GDK_PIXBUF_MODULE_FILE`:
+    The `librsvg` loaders.cache still contains all of the loaders from `gdk-pixbuf`. This might change in the future, so it is recommended that you explicitly include both.
+    As this change allows users to extend programs with their own loaders, it is recommended that you `--prefix` rather than `--set` `$GDK_PIXBUF_MODULE_FILE` with your values.
+    For example: `makeWrapper ... --prefix GDK_PIXBUF_MODULE_FILE : "${librsvg}/lib/gdk-pixbuf-2.0/2.10.0/loaders.cache:${gdk-pixbuf}/lib/gdk-pixbuf-2.0/2.10.0/loaders.cache"`
```
Maybe also note something like “But using `wrapGAppsHook`/`wrapGAppsNoGuiHook` with `extraGdkPixbufModules` is preferred.” to be safe.

jtojnar commented on pkgs/development/libraries/gdk-pixbuf/setup-hook.sh:
```diff
@@ -1,17 +0,0 @@
-findGdkPixbufLoaders() {
```
One use case that still is ¿not covered? is the use of loaders during `build`/`checkPhase`. Previously `findGdkPixbufLoaders` had us covered but now we need to set the environment variable manually. But maybe it is okay?

cwyc replied:
If we set GDK_PIXBUF_MODULE_FILE for the user, then should we also return to the behaviour of wrapGAppsHook setting it based on the build-time environment variable, so that what works in the build environment also works in the final package?

cwyc replied:
Also, would not being able to control order be an issue?

jtojnar replied:
I can imagine some use cases for wanting specific order but they probably be rare so I would not bother with that. People can always juggle with the variable contents if they need it.

Though I am split on having a separate attribute vs. using `buildInputs`. On one hand, separate attribute offers a finer control and the behaviour is more clear by the fact of being explicit. On the other hand all other wrappers just pick stuff up from inputs so not doing that for loaders might be confusing (largely mitigated by docs?) and deciding which dependency to put where might be confusing as well. :woman_shrugging:

cwyc committed:
updated documentation
will squash before merge

cwyc committed:
expanded package tests
found issue: there should be a better way to report that a loader is invalid
will be squashed before merge

cwyc committed:
cleaned up setuphook test
will squash before merge

cwyc commented on pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix:
```diff
@@ -64,6 +68,33 @@ makeSetupHook {
+        dontUnpack = true;
+        buildPhase = ''
+          mkdir -p $out
+          mkdir -p $lib/$(dirname ${gdk-pixbuf.cacheFile})
+          touch $lib/${gdk-pixbuf.cacheFile}
```
I think they should be separate, since the sample-project is wrapped by the setup hook and the mock loader is an input for the wrapper.

jtojnar replied:
Sorry, I meant to move it to the `Makefile`, then you will be able to use `src` & `installFlags`.

cwyc commented on pkgs/build-support/setup-hooks/wrap-gapps-hook/default.nix:
```diff
@@ -64,6 +68,33 @@ makeSetupHook {
+        ${expectSomeLineContainingYInFileXToMentionZ "${tested}/bin/foo" "GDK_PIXBUF_MODULE_FILE" "${mock-pixbuf-module}/${gdk-pixbuf.cacheFile}"}
+        ${expectSomeLineContainingYInFileXToMentionZ "${tested}/libexec/bar" "GDK_PIXBUF_MODULE_FILE" "${gdk-pixbuf}/${gdk-pixbuf.cacheFile}"}
+        ${expectSomeLineContainingYInFileXToMentionZ "${tested}/libexec/bar" "GDK_PIXBUF_MODULE_FILE" "${librsvg}/${gdk-pixbuf.cacheFile}"}
+        ${expectSomeLineContainingYInFileXToMentionZ "${tested}/libexec/bar" "GDK_PIXBUF_MODULE_FILE" "${mock-pixbuf-module}/${gdk-pixbuf.cacheFile}"}
```
I think I was wrong to put the mock loader it in $lib, I've changed it to $out. But the checks passed, brings up the issue that the wrapper doesn't check for the cache file's existence.

lovesegfault commented:
@cwyc Could you resolve the conflicts?

jtojnar commented:
Sorry I have not managed to review this early enough for this to make it into 2022.11. I have created a workaround for now  https://github.com/NixOS/nixpkgs/pull/153275
